### PR TITLE
refactor(rules): config structure

### PR DESCRIPTION
**Pull request by cmdoret**
## Proposed Changes

Changes the rule specification format to match the proposal in https://github.com/sdsc-ordes/tripsu/issues/45.
* adapt matching logic and associated tests, simplifying where possible.
* move invert cli flag into config (single source of truth for pseudonymization rules)
* move top-level function `match_rules` into the rules module
* mark fields as optional in serde so that users don't need to add all empty fields.

The mask and triple were cloned a few times by the matching rules, the refactored version passes everything by reference, resulting in a 13% speedup (15.9s -> 14.1s on a 2.3M triples file). That's ~164k triples/second.

## Types of Changes

What types of changes does your code introduce? _Put an `x` in the boxes that
apply_

- [ ] A **bug fix** (non-breaking change which fixes an issue).
- [x] A new **feature** (non-breaking change which adds functionality).
- [x] A **breaking change** (fix or feature that would cause existing
      functionality to not work as expected).
- [ ] A **non-productive** update (documentation, tooling, etc. if none of the
      other choices apply).

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask. We're here to
help! This is simply a reminder of what we are going to look for before merging
your code._

- [x] I have read the
      [CONTRIBUTING](https://github.com/sdsc-ordes/tripsu/blob/master/CONTRIBUTING.md)
      guidelines.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added the necessary documentation (if appropriate).

### CODE REVIEW

File: src/rules.rs [hunk 1]
```diff
@@ -4,238 +4,233 @@ use serde::{Deserialize, Serialize};
...
+                match_type_predicate(&n.iri, &triple.predicate.iri, type_map, rules)
+            }
         }
-        _ => return mask,
-    }
-}
-
-pub fn match_predicate_rule(predicate: &NamedNode, mask: TripleMask, rules: &Rules) -> TripleMask {
-    let NamedNode { iri: i } = predicate;
+        _ => false,
```
Comment (supermaxiste) [inline src/rules.rs:96]: issue(bug): sneaky missing `Literal` type. When running `tripsu` on the example data and config, "accessCode" values are not pseudonymized. The rules all work as expected and after a bit of investigation I landed on this line. I think that for best practices, it's better to be explicit on the types when using `match` and avoid using `_`.

File: src/rules.rs [hunk 2]
```diff
@@ -4,238 +4,233 @@ use serde::{Deserialize, Serialize};
...
 ) -> TripleMask {
-    match subject {
-        Subject::NamedNode(n) => {
-            return mask | match_type_rule_named_node(true, n, mask, rules, type_map);
-        }
-        Subject::BlankNode(_) => return mask,
-    }
+    let pseudo_subject = match &triple.subject {
+        Subject::NamedNode(n) => match_type(&n.iri, rules, type_map),
+        _ => false,
```
Comment (supermaxiste) [inline src/rules.rs:64]: suggestion(non-blocking): see comment on L96 about explicit typing when using `match`

File: src/rules.rs [hunk 3]
```diff
@@ -4,238 +4,233 @@ use serde::{Deserialize, Serialize};
...
+    };
+    let pseudo_object = match &triple.object {
+        Term::NamedNode(n) => match_type(&n.iri, rules, type_map),
+        _ => false,
```
Comment (supermaxiste) [inline src/rules.rs:68]: suggestion(non-blocking): see comment on L96 about explicit typing when using match

**Review by supermaxiste — CHANGES_REQUESTED** [review]
Very well done @cmdoret 🥳 

**praise**: the code for rules looks both more structured and cleaner
**praise**: the new config is definitely easier to understand and use

I only have a major comment and an open question related to it:

**issue**: Since we're playing with the rules, I tried a test run on my end after compiling your code and noticed that all literal values were not pseudonymized in the output. After a bit of digging I stumbled upon a matching condition that was looking for only `NamedNode` as a type and used `_` for any other case. I suggested avoiding this when possible because it goes a bit against Rust's principle of explicit typing, let me know your thoughts.

```
./target/debug/tripsu pseudo --index tests/data/type_map.nt --config tests/data/config.yaml tests/data/test.nt
 INFO Args: PseudoArgs { index: "tests/data/type_map.nt", input: "tests/data/test.nt", config: "tests/data/config.yaml", output: "-", secret: None }
<http://example.org/4cbe42b24189cb18b362a8cd0e952c482ad3b757ae96dc2590ba8b94d0911895> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://xmlns.com/foaf/0.1/Person> .
<http://example.org/4cbe42b24189cb18b362a8cd0e952c482ad3b757ae96dc2590ba8b94d0911895> <http://xmlns.com/foaf/0.1/holdsAccount> <http://example.org/a7ee682dbe9f90d4b52c156744757a4aae47df7da05f2ac91261fb4c66cb5340> .
<http://example.org/a7ee682dbe9f90d4b52c156744757a4aae47df7da05f2ac91261fb4c66cb5340> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://xmlns.com/foaf/OnlineAccount> .
<http://example.org/a7ee682dbe9f90d4b52c156744757a4aae47df7da05f2ac91261fb4c66cb5340> <http://schema.org/name> "my_account32" . <- should be pseudo
<http://example.org/a7ee682dbe9f90d4b52c156744757a4aae47df7da05f2ac91261fb4c66cb5340> <http://schema.org/accessCode> "secret-123" . <- should be pseudo
<http://example.org/4cbe42b24189cb18b362a8cd0e952c482ad3b757ae96dc2590ba8b94d0911895> <http://schema.org/name> "Alice" . <- should be pseudo
[...]
```

**suggestion(tests)**: the bug above is very funny because all unit tests pass, meaning that all of our main functions work. The issue arises when you run the whole pipeline and a "side" function doesn't do everything it's supposed to. How can we address this? 
I suggest adding a test function in `pass_second` to test the encryption of several triples using `rstest`. Let me know if you have a better idea 👍

File: src/rules.rs (hunk 1 above)
Comment (cmdoret) [reply src/rules.rs:96]: i.e. the fix in [544fceb](https://github.com/sdsc-ordes/tripsu/pull/47/commits/544fceb54f507b42c437c923069ee57cefac7eb9), but I'll add a test in second_pass as you suggested

File: src/rules.rs (hunk 3 above)
Comment (cmdoret) [reply src/rules.rs]: fixed in [725595f](https://github.com/sdsc-ordes/tripsu/pull/47/commits/725595f04533e5aab05ad445f830e81c0b474e8d)

File: src/rules.rs (hunk 2 above)
Comment (cmdoret) [reply src/rules.rs]: fixed in 725595f

**Review by supermaxiste — APPROVED** [review]
Apologies for pointing out the wrong line for the bug and excellent job with fixing and the tests 😎 

**praise**: seeing the `index!` macro in action with multiple entries is so clean! 
**praise**: the test for triples is exactly how I imagined it 
**praise**: thank you for renaming pass_first and pass_second, now it feels like we're 100% in release state

All looks good to me!

**Pull request merged**